File: apps/jsx/src/mrly/two/serializer.ts
import { Cell2d } from "./models";
import type { RGBA } from "../colors";

export type CellDict2d = {
  width: number;
  height: number;
  types?: number[][];
  colors?: number[][][];
  tags?: number[][];
};

// 2D

export function toDict2d(cell: Cell2d): CellDict2d {
  const data: CellDict2d = { width: cell.width, height: cell.height };
  if (cell._types !== null) data.types = cell._types;
  if (cell._colors !== null) data.colors = cell._colors.map((row) => row.map((c) => (c ? [...c] : [0, 0, 0, 0])));
  if (cell._tags !== null) data.tags = cell._tags;
  return data;
}

export function fromDict2d(data: CellDict2d): Cell2d {
  return new Cell2d({
    width: data.width,
    height: data.height,
    types: data.types ?? null,
    colors: data.colors ? (data.colors.map((row) => row.map((c) => c as RGBA)) as RGBA[][]) : null,
    tags: data.tags ?? null,
  });
}

export function toList2d(cell: Cell2d): number[][] {
  return cell.types.map((row) => row.slice());
}

export function fromList2d(data: number[][]): Cell2d {
  return new Cell2d({ types: data });
}

export function toStrings2d(cell: Cell2d): string[] {
  return cell.types.map((row) => row.join(""));
}

export function fromStrings2d(data: string[]): Cell2d {
  const types = data.map((row) => [...row].map(Number));
  return new Cell2d({ types });
}

File: crates/mrlymath/src/two/serializer.rs
use super::models::Cell2d;
use crate::dim::serializer::{byte_grid, color_grid, count_grid, parse, tag_layer, types_field};
use mrlycore::errors::{value_error, Result};
use mrlycore::json;
use mrlycore::tensor::Tensor;

/// Returns the cell's types as rows of bytes.
fn to_lists(cell: &Cell2d) -> Vec<Vec<u8>> {
    let (h, w) = (cell.height(), cell.width());
    (0..h)
        .map(|y| (0..w).map(|x| cell.types().get(&[y, x])).collect())
        .collect()
}

/// Builds a cell from rows of bytes, or an error when the rows are empty or ragged.
fn from_lists(lists: &[Vec<u8>]) -> Result<Cell2d> {
    if lists.is_empty() {
        return value_error("cannot build a cell from an empty list.");
    }
    let (h, w) = (lists.len(), lists[0].len());
    if lists.iter().any(|row| row.len() != w) {
        return value_error("all rows must have the same length.");
    }
    let data: Vec<u8> = lists.iter().flatten().copied().collect();
    Ok(Cell2d::new(Tensor::of(data, vec![h, w])))
}

/// Serializes the cell to a JSON string of its types, with colors and tags when present.
pub fn to_json(cell: &Cell2d) -> String {
    let mut data = json!({
        "v": 1,
        "width": cell.width(),
        "height": cell.height(),
        "types": to_lists(cell),
    });
    if let Some(colors) = &cell.cell.colors {
        data["colors"] = json!(colors.chunks(cell.width()).collect::<Vec<_>>());
    }
    if let Some(tags) = &cell.cell.tags {
        let (h, w) = (cell.height(), cell.width());
        let nested: Vec<Vec<i64>> = (0..h)
            .map(|r| (0..w).map(|c| tags.at(r * w + c)).collect())
            .collect();
        data["tags"] = json!(nested);
    }
    data.to_string()
}

/// Restores a cell from its JSON string, colors and tags included, or a parse error.
pub fn from_json(text: &str) -> Result<Cell2d> {
    let data = parse(text)?;
    let lists = byte_grid(types_field(&data)?)?;
    let mut cell = from_lists(&lists)?;
    if let Some(colors) = data.get("colors") {
        let nested = color_grid(colors)?;
        cell.cell.colors = Some(nested.into_iter().flatten().collect());
    }
    if let Some(tags) = data.get("tags") {
        let shape = vec![cell.height(), cell.width()];
        cell.cell.tags = Some(tag_layer(&count_grid(tags)?, shape)?);
    }
    Ok(cell)
}
